File: Data Structures & Algorithms/binary-tree-zigzag-level-order-traversal/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        
        if not root:
            return []
        
        q = deque()
        q.append(root)
        res = []

        while q:
            deck = []
            for i in range(len(q)):
                node = q.popleft()
                deck.append(node.val)
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
            res.append(deck)

        zigRes = []
        for i in range(len(res)):
            if i % 2 == 0:
                zigRes.append(res[i])
            else:
                zigRes.append(res[i][::-1])
        return zigRes

File: Data Structures & Algorithms/palindromic-substrings/submission-0.java
class Solution {
    public int countSubstrings(String s) {
        
        int res = 0;

        for ( int i = 0; i < s.length(); i++ ) {
            res += hasPalindromes(s,i,i);
            res += hasPalindromes(s,i,i+1);
        }
        return res;
    }

    private int hasPalindromes(String s, int left, int right) {

        // if ( s.length() <= 1 ) return true;
        // int middle = s.length() / 2;
        // int left   = middle - 1;
        // int right  = middle + (s.length() % 2) ; // Odd or Even

        int res = 0;
        while ( left >= 0 && right < s.length() && s.charAt(left) == s.charAt(right) ) {
            left--;
            right++;
            res++;
        }

        return res; 


    }
}
